### PR TITLE
HdMapUtils refactor (PR 2/6) - extend lanelet_wrapper: add ::distance and necessary parts of ::route, ::lanelet_map, ::traffic_lights

### PR DESCRIPTION
**Pull request by dmoszynski**
# Description

## Abstract

This is the second PR of 6 related to the `HdMapUtils` refactor (for the first one, see #1472).
For a full overview of changes that all 6 PRs will introduce, please see the [Abstract of the first PR](https://github.com/tier4/scenario_simulator_v2/pull/1472#issue-2715734695).

## Background

The `lanelet_wrapper` namespace contains:
- [`LaneletLoader`](https://github.com/tier4/scenario_simulator_v2/blob/ref/RJD-1387-hdmap-utils-to-lanelet-wrapper-distance/simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_loader.hpp)
- [`LaneletWrapper`](https://github.com/tier4/scenario_simulator_v2/blob/ref/RJD-1387-hdmap-utils-to-lanelet-wrapper-distance/simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp)
-  [`RouteCache`,`CenterPointsCache`, `LaneletLengthCache`](https://github.com/tier4/scenario_simulator_v2/blob/ref/RJD-1387-hdmap-utils-to-lanelet-wrapper-distance/simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp)
- [`TrafficRulesWithRoutingGraph`](https://github.com/tier4/scenario_simulator_v2/blob/ref/RJD-1387-hdmap-utils-to-lanelet-wrapper-distance/simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp)
- 6 category-specific sub-namespaces

These static methods are used by the sub-namespaces of `lanelet_wrapper`.
`lanelet_wrapper` namespace consist of 6 sub-namespaces:
- [`lanelet_wrapper::lanelet_map`][lanelet_wrapper_lanelet_map],
- `lanelet_wrapper::pose`,
- [`lanelet_wrapper::distance`][lanelet_wrapper_distance],
- [`lanelet_wrapper::route`][lanelet_wrapper_route],
- `lanelet_wrapper::lane_change`,
- [`lanelet_wrapper::traffic_lights`][lanelet_wrapper_traffic_lights].

[lanelet_wrapper_lanelet_map]: https://github.com/tier4/scenario_simulator_v2/pull/1478/files#diff-039b6468c0615f08684b3ef6b3ee8da11838ca9f293373f243256a144831e921
[lanelet_wrapper_distance]: https://github.com/tier4/scenario_simulator_v2/pull/1478/files#diff-43d777207527ae34a6fe956997f3ba04ed172d67db836aff7be0fc78235d0e00
[lanelet_wrapper_route]: https://github.com/tier4/scenario_simulator_v2/pull/1478/files#diff-13a5da84a23deb82bfb4395e341203e520b7079b24c8e13bdaf892a359fd9e28
[lanelet_wrapper_traffic_lights]: https://github.com/tier4/scenario_simulator_v2/pull/1478/files#diff-3199c2e75265441a9dbff6ebe9aa7034f76ca75271525c566dc08224153ec944

As a result of the merge of this PR 2/6, the namespaces [`lanelet_wrapper::lanelet_map`][lanelet_wrapper_lanelet_map], [`lanelet_wrapper::distance`][lanelet_wrapper_distance], [`lanelet_wrapper::route`][lanelet_wrapper_route] and [`lanelet_wrapper::traffic_lights`][lanelet_wrapper_traffic_lights] were added to the project. But most importantly, the [`::distance`](https://github.com/tier4/scenario_simulator_v2/pull/1478/files#diff-cfdae9768cc43e77c1a48480ec01c84647998a588a515316ce82a330a2cbc0d1) from `utils` is adapted for its use.
In addition to this, PR is making several other improvements to the code, see below for details.

## Details

The `distance` namespace in `lanelet_wrapper` and only the strictly necessary parts of `route` and `traffic_lights` are developed.

> [!IMPORTANT]
> By moving from `HdMapUtils`, we mean moving the entire member function to `lanelet_wrapper` - making it a free function.
> This is accessible to `traffic_simulator` through free functions in `utils` namespace.
>
> Before these changes the structure looked like this:
>
> `<HdMapUtils member function name>`
>
> After the changes structure looks like this:
>
> `<utils namespace and function name>` &rarr; `<lanelet_wrapper namespace and function name>`

**Below is the list of moved function (member functions transformed to free functions)**

> [!TIP]
> **Most** functions that have been removed from `HdMapUtils` have been renamed in such a way that the `get` prefix has been removed and thus the first letter of the new function name has been made lower case.
>
> Example of such name change can be the following
>
> `HdMapUtils::getLateralDistance` &rarr; `lanelet_wrapper::distance::lateralDistance`

| Old name | New name |
| --- | --- |
| `HdMapUtils::getLateralDistance` | `distance::lateralDistance` &rarr; `lanelet_wrapper::distance::lateralDistance` |
| `HdMapUtils::getLongitudinalDistance` | `distance::longitudinalDistance` &rarr; `lanelet_wrapper::distance::longitudinalDistance` |
| `HdMapUtils::getDistanceToStopLine` | `distance::distanceToStopLine` &rarr; `lanelet_wrapper::distance::distanceToStopLine` |
| `HdMapUtils::getDistanceToTrafficLightStopLine` | `distance::distanceToTrafficLightStopLine` &rarr; `lanelet_wrapper::distance::distanceToTrafficLightStopLine` |
| `HdMapUtils::getDistanceToCrosswalk` | `distance::distanceToCrosswalk` &rarr; `lanelet_wrapper::distance::distanceToCrosswalk` |
| `HdMapUtils::getLeftBound` | `lanelet_wrapper::lanelet_map::leftBound` |
| `HdMapUtils::getRightBound` | `lanelet_wrapper::lanelet_map::rightBound` |
| `HdMapUtils::getLaneletPolygon` | `lanelet_wrapper::lanelet_map::laneletPolygon` |
| `HdMapUtils::getStopLinePolygon` | `lanelet_wrapper::lanelet_map::stopLinePolygon` |
| `HdMapUtils::toPolygon` | `lanelet_wrapper::lanelet_map::toPolygon` |
| `HdMapUtils::getRightOfWayLaneletIds` | `lanelet_wrapper::lanelet_map::rightOfWayLaneletIds` |
| `HdMapUtils::getTrafficSignRegulatoryElementsOnPath` | `lanelet_wrapper::lanelet_map::trafficSignsOnPath` |
| `HdMapUtils::getStopLinesOnPath` | `lanelet_wrapper::lanelet_map::stopLinesOnPath` |
| `HdMapUtils::getStopLineIdsOnPath` | `lanelet_wrapper::lanelet_map::stopLineIdsOnPath` |
| `HdMapUtils::getRoute` | `route::route` &rarr; `lanelet_wrapper::route::route` |
| `HdMapUtils::getTrafficLightStopLinesPoints` | `lanelet_wrapper::traffic_lights::trafficLightStopLinesPoints` |
| `HdMapUtils::getTrafficLightIdsOnPath` | `lanelet_wrapper::traffic_lights::trafficLightStopLinesPoints` |
| `HdMapUtils::getTrafficLights` | `lanelet_wrapper::traffic_lights::toAutowareTrafficLights` |
| `HdMapUtils::getTrafficLightRegulatoryElementsOnPath` | `lanelet_wrapper::traffic_lights::autowareTrafficLightsOnPath` |

## References

# Destructive Changes

# Known Limitations

### CODE REVIEW

**Comment by TauTheLepton** [comment]
No regressions confirmed [here](https://github.com/tier4/sim_evaluation_tools/issues/451#issuecomment-2623789122)

File: simulation/traffic_simulator/src/utils/distance.cpp [hunk 1]
```diff
@@ -110,28 +109,24 @@ auto longitudinalDistance(
...
       for (const auto & to_pose : to_poses) {
-        if (
-          const auto distance = longitudinalDistance(
-            CanonicalizedLaneletPose(from_pose), CanonicalizedLaneletPose(to_pose), false,
-            include_opposite_direction, routing_configuration, hdmap_utils_ptr)) {
-          distances.emplace_back(distance.value());
+        if (const auto distance = distance::longitudinalDistance(
+              CanonicalizedLaneletPose(from_pose), CanonicalizedLaneletPose(to_pose), false,
+              include_opposite_direction, routing_configuration);
+            distance.has_value() and
```
Comment (hakuturu583) [inline simulation/traffic_simulator/src/utils/distance.cpp:118]: In lines 115~117.

````
        if (const auto distance = distance::longitudinalDistance(
              CanonicalizedLaneletPose(from_pose), CanonicalizedLaneletPose(to_pose), false, false
              include_opposite_direction, routing_configuration);
````

This line is always True, so it does not seem to make sense to add it to the and condition.

File: simulation/traffic_simulator/src/utils/pose.cpp [hunk 2]
```diff
@@ -284,19 +284,17 @@ auto boundingBoxRelativePose(
 /// @todo HdMapUtils will be removed when lanelet_wrapper::distance is added
 auto relativeLaneletPose(
   const CanonicalizedLaneletPose & from, const CanonicalizedLaneletPose & to,
-  const RoutingConfiguration & routing_configuration,
-  const std::shared_ptr<hdmap_utils::HdMapUtils> & hdmap_utils_ptr) -> LaneletPose
+  const RoutingConfiguration & routing_configuration) -> LaneletPose
 {
   constexpr bool include_adjacent_lanelet{false};
-  constexpr bool include_opposite_direction{true};
+  constexpr bool include_opposite_direction{false};
```
Comment (hakuturu583) [inline simulation/traffic_simulator/src/utils/pose.cpp:290]: Please leave comment why the value boolean was inverted.

File: simulation/traffic_simulator/src/utils/distance.cpp (hunk 1 above)
Comment (TauTheLepton) [reply simulation/traffic_simulator/src/utils/distance.cpp:118]: I do not think this will always be true. The function `longitudinalDistance` can sometimes return an empty optional.

What is more, this part of the check has not been changed by this PR. The check whether the distance returned by `longitudinalDistance` function contains a value has been implemented before (see [here](https://github.com/tier4/scenario_simulator_v2/blob/7b3dba675c900d79a3927603847ccd80bab7f17c/simulation/traffic_simulator/src/utils/distance.cpp#L117-L119)). The only change that has been made to this portion of the code is adding the additional conditions that are in lines 119-120.

Please note that the code
```cpp
        if (const auto distance = distance::longitudinalDistance(
              CanonicalizedLaneletPose(from_pose), CanonicalizedLaneletPose(to_pose), false,
              include_opposite_direction, routing_configuration);
            distance.has_value() and
            (not min_distance.has_value() or
             (std::abs(distance.value()) < std::abs(min_distance.value())))) {
          // something
        }
```
is functionally equivalent to
```cpp
        if (
          const auto distance = distance::longitudinalDistance(
            CanonicalizedLaneletPose(from_pose), CanonicalizedLaneletPose(to_pose), false,
            include_opposite_direction, routing_configuration)) {
          if (
            not min_distance.has_value() or
            (std::abs(distance.value()) < std::abs(min_distance.value()))) {
            // something
          }
        }
```

Which would be just an additional check performed inside the [existing if check](https://github.com/tier4/scenario_simulator_v2/blob/7b3dba675c900d79a3927603847ccd80bab7f17c/simulation/traffic_simulator/src/utils/distance.cpp#L117-L119).

The first version was implemented in order not to complicate the code unnecessarily.

File: simulation/traffic_simulator/src/utils/pose.cpp (hunk 2 above)
Comment (TauTheLepton) [reply simulation/traffic_simulator/src/utils/pose.cpp:290]: This value was inverted, because of this [PR](https://github.com/tier4/scenario_simulator_v2/pull/1468).

It introduced the change where `getLeftLaneletIds` and `getRightLaneletIds` was throwing error when called with `include_opposite_direction = true` (see [here](https://github.com/tier4/scenario_simulator_v2/commit/e6f8c24f614b1a2236b0ee31de51f2a6748e8fec)).

The functions from that PR are `lanelet_wrapper::pose::leftLaneletIds` and `lanelet_wrapper::pose::rightLaneletIds` [here](https://github.com/tier4/scenario_simulator_v2/blob/773a3802dc3926355c3ee2d9b321cf181f4c56c9/simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp#L475-L502).

These functions are used by `lanelet_wrapper::pose::toLaneletPoses` [here](https://github.com/tier4/scenario_simulator_v2/blob/773a3802dc3926355c3ee2d9b321cf181f4c56c9/simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp#L191-L210), which is used in `distance::longitudinalDistance` [here](https://github.com/tier4/scenario_simulator_v2/blob/773a3802dc3926355c3ee2d9b321cf181f4c56c9/simulation/traffic_simulator/src/utils/distance.cpp#L102-L110).

---

So to sum this up, the call stack looks like this:
- `pose::relativeLaneletPose`
- `distance::longitudinalDistance`
- `lanelet_wrapper::pose::toLaneletPoses`
- `lanelet_wrapper::pose::leftLaneletIds` and `lanelet_wrapper::pose::rightLaneletIds`

The `include_opposite_direction` parameter is propagated through all these functions, and the last ones (lanelet ids) throw error when the parameter is true.

File: simulation/traffic_simulator/src/utils/distance.cpp (hunk 1 above)
Comment (hakuturu583) [reply simulation/traffic_simulator/src/utils/distance.cpp:118]: > I do not think this will always be true.

In `const auto distance = distance::longitudinalDistance,` distance is not always true, that's right.

But, your code check the distance is not std::nullopt by using if section.

```
        if (const auto distance = distance::longitudinalDistance(
              CanonicalizedLaneletPose(from_pose), CanonicalizedLaneletPose(to_pose), false,
              include_opposite_direction, routing_configuration);
```

And after this,

```
distance.has_value()
```

If the distance is not std::nullopt, distance.has_value() = true.
If the distance is std::nullopt, distance.has_value() = false.

The two expressions are connected by the and operator, but this does not make sense because the checks are the same. The code is unnecessarily complicated and unnecessary conditions should be removed.

File: simulation/traffic_simulator/src/utils/pose.cpp (hunk 2 above)
Comment (hakuturu583) [reply simulation/traffic_simulator/src/utils/pose.cpp:290]: Please put the above information in the comments of the source code instead of putting it in Discussion on GitHub.

**Review by hakuturu583 — CHANGES_REQUESTED** [review]

File: simulation/traffic_simulator/src/utils/distance.cpp (hunk 1 above)
Comment (TauTheLepton) [reply simulation/traffic_simulator/src/utils/distance.cpp:118]: # If statement *init-statement*
As it is noted on the [cppreference page](https://en.cppreference.com/w/cpp/language/if) - if statement consists just *condition* or *init-statement* and *condition* separated by a semicolon. 

The *condition* can be a declaration. In such case, the declared variable is converted to `bool` and used as a decision variable.

However, when both *init-statement* and *condition* are used, the *init-statement* is not evaluated. The decision variable is still in the *condition* part.
If *init-statement* it used to initialize the variable, it narrows down the scope of the variable.
Screenshot from 2025-04-29 13-19-42

So if we want to check both the value initialized in *init-statement* and some other condition, we need to explicitly add the value check to the condition part of the if statement.
See the example below where in both cases I initialize the value with `false`.
Only in the second statement when I explicitly added the value check did the if statement yield false.
Screenshot from 2025-04-29 13-32-44

# Our case
So in the case with `std::optional`, when the value `distance` was declared in *condition* part it was checked (using bool operator implicitly).
However, when I have added additional checks, the `distance` declaration was moved to *init-statement*, and so I had to include an explicit check in the *condition* part so that the behavior stayed the same - the `distance` value was checked.

File: simulation/traffic_simulator/src/utils/pose.cpp (hunk 2 above)
Comment (hakuturu583) [reply simulation/traffic_simulator/src/utils/pose.cpp:290]: Please do not leave these comment on GitHub.
Please leave comments on source code.
It is necessary for checking why the value was changed.

**Review by hakuturu583 — APPROVED** [review]

**Review by hakuturu583 — CHANGES_REQUESTED** [review]
I confirmed regression.

https://github.com/tier4/sim_evaluation_tools/issues/555

**Comment by hakuturu583** [comment]
@TauTheLepton This PR was degraded.
Please check and fix problems.

**Comment by hakuturu583** [comment]
@TauTheLepton I will conform and fix this pull request.

**Comment by TauTheLepton** [comment]
No regressions confirmed [here](https://github.com/tier4/sim_evaluation_tools/issues/555#issuecomment-3466657402).

**Review by HansRobo — APPROVED** [review]
